Userland/Libraries/LibJS/AST.h

```diff
@@ -97,6 +97,45 @@ class ASTNode : public RefCounted<ASTNode> {
     u32 m_end_offset { 0 };
 };
 
+// This is a helper class that packs an array of T after the AST node, all in the same allocation.
+template<typename Derived, typename Base, typename T>
+class ASTNodeWithTailArray : public Base {
+public:
+    template<typename... Args>
+    static NonnullRefPtr<Derived> create(size_t tail_size, SourceRange source_range, Args&&... args)
+    {
+        static_assert(alignof(Derived) % alignof(T) == 0, "need padding for tail array");
+        auto memory = ::operator new(sizeof(Derived) + tail_size * sizeof(T));
+        return adopt_ref(*::new (memory) Derived(move(source_range), static_cast<decltype(args)>(args)...));
```

Comment:
The `static_cast` is my way of spelling `std::forward`, you might want to change that for consistency.